src/Converter.cc

```diff
@@ -229,13 +240,270 @@ void Converter::ConvertImpl(tinyxml2::XMLElement *_elem,
+
+    std::string elemName = elem->Name();
+
+    // skip element if not one of the following
+    if (elemName != "frame" && elemName != "joint"
+        && elemName != "link" && elemName != "model")
+    {
+      elem = nextElem;
+      continue;
+    }
+
+    std::string attrName = elem->Attribute("name");
+
+    size_t found = attrName.find("::");
+    if (found == std::string::npos)
+    {
+      // recursive unnest
+      if (elemName == "model")
+      {
+        // std::cout << "before recursive:\n"
+        //           << PrintElement(elem) << std::endl;
+        Unnest(elem);
+
+        // std::cout << "unnested model:\n" << PrintElement(elem) << std::endl;
+        _elem->InsertEndChild(elem);
+        break;
+      }
+
+      elem = nextElem;
+      continue;
+    }
+
+    std::string newModelName = attrName.substr(0, found);
+    tinyxml2::XMLElement *newModel = doc->NewElement("model");
+    newModel->SetAttribute("name", newModelName.c_str());
+
+    Converter::TupleVector deleteElems =
+        FindNewModelElements(doc,
+                            copy->ToElement(),
+                            newModel,
+                            found + 2);
+
+    // std::cout << "newModel:\n" << PrintElement(newModel) << std::endl;
+
+    if (!deleteElems.empty())
+    {
+      _elem->InsertEndChild(newModel);
+
+      // delete unnested elements
+      for (std::tuple<std::string, std::string> deleteElem : deleteElems)
+      {
+        tinyxml2::XMLElement *e =
+          _elem->FirstChildElement(std::get<0>(deleteElem).c_str());
+
+        std::string eName = std::get<1>(deleteElem);
+
+        while (eName != std::string(e->Attribute("name")))
+            e = e->NextSiblingElement(std::get<0>(deleteElem).c_str());
+
+        if (std::string(e->Attribute("name")) == eName)
+          _elem->DeleteChild(e);
+      }
+
+      nextElem = _elem->FirstChildElement();
+    }
+
+    elem = nextElem;
+  }
+}
+
+/////////////////////////////////////////////////
+Converter::TupleVector Converter::FindNewModelElements(
+                                                tinyxml2::XMLDocument *_doc,
+                                                tinyxml2::XMLElement *_copy,
+                                                tinyxml2::XMLElement *_newModel,
+                                                const size_t &_newNameIdx)
+{
+  Converter::TupleVector deleteElems;
+  std::string newModelName = _newModel->Attribute("name");
+
+  tinyxml2::XMLElement *elem = _copy->FirstChildElement(), *nextElem = nullptr;
+  while (elem)
+  {
+    nextElem = elem->NextSiblingElement();
+
+    std::string elemName = elem->Name();
+    std::string elemAttrName;
+
+    if (elem->Attribute("name"))
+      elemAttrName = elem->Attribute("name");
+
+    if (elemAttrName.empty() ||
+        elemAttrName.find(newModelName) == std::string::npos ||
+        (elemName != "frame" && elemName != "joint" &&
+        elemName != "link" && elemName != "model"))
```

Comment:
The spec also has `//model/gripper`. Although I've never seen it used and it's not supported in ign-gazebo, I think we should handle it for good measure.

Reply:
8cd7f4d876fa787f297eba62a30f8e83e5f03721